packages/cli/src/__tests__/commands/virtualdevices/events.test.ts

```diff
@@ -0,0 +1,370 @@
+import VirtualDeviceEventsCommand from '../../../commands/virtualdevices/events'
+import { CustomCapabilityStatus } from '@smartthings/core-sdk/dist/endpoint/capabilities'
+import {
+	chooseAttribute,
+	chooseCapability,
+	chooseComponent,
+	chooseUnit,
+	chooseValue,
+} from '../../../lib/commands/virtualdevices/virtualdevices-util'
+
+
+jest.mock('../../../lib/commands/virtualdevices/virtualdevices-util')
+
+describe('VirtualDeviceEventsCommand', () => {
+	const mockSelectFromList = jest.mocked(selectFromList)
+
+	const mockInputAndOutputItem = jest.mocked(inputAndOutputItem)
+	const createEventsSpy = jest.spyOn(VirtualDevicesEndpoint.prototype, 'createEvents').mockImplementation()
+	const getCapabilitySpy = jest.spyOn(CapabilitiesEndpoint.prototype, 'get')
+	const getDeviceSpy = jest.spyOn(DevicesEndpoint.prototype, 'get')
+
+	it('calls correct endpoint', async () => {
+		const createRequest: DeviceEvent[] = [
+			{
+				component: 'main',
+				capability: 'switch',
+				attribute: 'switch',
+				value: 'on',
+			},
+		]
+
+		mockInputAndOutputItem.mockImplementationOnce(async (_command, _config, actionFunction) => {
+			await actionFunction(undefined, createRequest)
+		})
+
+		mockSelectFromList.mockResolvedValueOnce('device-id')
+
+		await expect(VirtualDeviceEventsCommand.run(['device-id'])).resolves.not.toThrow()
+
+		expect(mockInputAndOutputItem).toBeCalledWith(
+			expect.any(VirtualDeviceEventsCommand),
+			expect.anything(),
+			expect.any(Function),
+			expect.anything(),
+		)
+
+		expect(mockSelectFromList).toHaveBeenCalledTimes(1)
+		expect(mockSelectFromList).toBeCalledWith(
+			expect.any(VirtualDeviceEventsCommand),
+			expect.objectContaining({
+				primaryKeyName: 'deviceId',
+				sortKeyName: 'label',
+			}),
+			expect.objectContaining({
+				preselectedId: 'device-id',
+			}),
+		)
+		expect(createEventsSpy).toHaveBeenCalledTimes(1)
+		expect(createEventsSpy).toBeCalledWith('device-id', createRequest)
+	})
+
+	it('string command line argument input', async () => {
+		const expectedCreateRequest: DeviceEvent[] = [
+			{
+				component: 'main',
+				capability: 'switch',
+				attribute: 'switch',
+				value: 'on',
+			},
+		]
+
+		mockInputAndOutputItem.mockImplementationOnce(async (_command, _config, actionFunction, inputProcessor) => {
+			const data = await inputProcessor.read()
+			await actionFunction(undefined, data)
+		})
+
+		mockSelectFromList.mockResolvedValueOnce('device-id')
+		getCapabilitySpy.mockResolvedValueOnce({
+			id: 'switch',
+			version: 1,
+			status: CustomCapabilityStatus.LIVE,
+			name: 'Switch',
+			attributes: {
+				switch: {
+					schema: {
+						type: 'object',
+						properties: {
+							value: {
+								title: 'IntegerPercent',
+								type: 'string',
+								enum: ['on', 'off'],
+							},
+						},
+						additionalProperties: false,
+						required: [
+							CapabilitySchemaPropertyName.VALUE,
+						],
+					},
+					enumCommands: [],
+				},
+			},
+			commands: {},
```

Comment:
Instead of including all properties, we have also been using a pattern of stubbing an object with only the properties required by the test case and casting the result to the required type. [Here is an example](https://github.com/SmartThingsCommunity/smartthings-cli/blob/237f6ebedb1acc4d39fb53123f23aa184ac87d8d/packages/cli/src/__tests__/commands/devices/update.test.ts#L37).

It makes the tests less verbose and easier to read.

Reply:
👍 , though for some reason I had to cast the shortened version to `unknown` before casting it to `Capability`. Not sure I understand what that wasn't necessary for the Device example

Reply:
This is expected. TypeScripts rules assertion rules are conservative and more complex objects will hit this. Documented [here](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions).